macosx/DispKbd.c

```diff
@@ -211,11 +208,98 @@ DisplayKbd_PollHost(ARMul_State *state)
   return 0;
 }
 
+#define SDD_Name(x) sdd_Mac_##x
+
+typedef uint32_t SDD_HostColour;
+typedef SDD_HostColour *SDD_Row;
+
+static void SDD_Name(Host_PollDisplay)(ARMul_State *state)
+{
+  RefreshMouse(state);
+  updateDisplay(0, 0, 800, 600, 1);
+}
+
+static void SDD_Name(Host_ChangeMode)(ARMul_State *state,int width,int height,
+                                      int hz);
+
+static SDD_HostColour SDD_Name(Host_GetColour)(ARMul_State *state,uint_fast16_t col)
+{
+  // TODO: CHECK THIS!
+  /* Convert to 5-bit component values */
+  int r = (col & 0x00f) << 1;
+  int g = (col & 0x0f0) >> 3;
+  int b = (col & 0xf00) >> 7;
+  /* May want to tweak this a bit at some point? */
+  r |= r>>4;
+  g |= g>>4;
+  b |= b>>4;
+  return (r<<24) | (g<<16) | (b<<8) | 0xFF;
+}
+
+static void SDD_Name(Host_SkipPixels)(ARMul_State *state,SDD_Row *row,
+                               unsigned int count)
+{
+  (*row) += count;
+}
+
+static SDD_Row SDD_Name(Host_BeginRow)(ARMul_State *state,int row,int offset)
+{
+  SDD_Row dibbmp = (unsigned int*)screenbmp;
+
+  return &dibbmp[(MonitorHeight-(row+1))*MonitorWidth+offset];
+}
+
+const int SDD_RowsAtOnce = 1;
+
+static void SDD_Name(Host_EndRow)(ARMul_State *state,SDD_Row *row)
+{
+  
+}
+
+static void SDD_Name(Host_WritePixel)(ARMul_State *state,SDD_Row *row,
+                               SDD_HostColour col)
+{
+  *(*row)++ = col;
+}
+
+static void SDD_Name(Host_WritePixels)(ARMul_State *state,SDD_Row *row,
+                                SDD_HostColour col,unsigned int count)
+{
+  while(count--) {
+    *(*row)++ = col;
+  }
+}
+
+static void SDD_Name(Host_BeginUpdate)(ARMul_State *state,SDD_Row *row,
+                                    unsigned int count)
+{
+  
+}
+
+static void SDD_Name(Host_EndUpdate)(ARMul_State *state,SDD_Row *row)
+{
+  
+}
+
+#include "../arch/stddisplaydev.c"
+
+static void SDD_Name(Host_ChangeMode)(ARMul_State *state,int width,int height,
+                                      int hz)
+{
+  HD.Width = width;
+  HD.Height = height;
+  HD.XScale = 1;
+  HD.YScale = 1;
+  resizeWindow(HD.Width,HD.Height);
```

Comment:
This is being called from the emulator thread, but it uses APIs that must be called from the main thread, leading to a crash.